Debian/Ubuntu: use non-root evcc user

pauxus opened the pull request:
- include preinstall script that stops the service (with flag) and
  copies  over the database
- postinstall also recognizes restart flag in above case
- unit file uses evcc user

Fixes #4900

Refs https://github.com/evcc-io/evcc/pull/4878

pauxus committed:
Use distinct user for evcc on debian/ubuntu

- include preinstall script that stops the service (with flag) and
  copies  over the database
- postinstall also recognizes restart flag in above case
- unit file uses evcc user

Fixes #4900

pauxus committed:
Style, chown for copied db

andig commented on packaging/scripts/preinstall.sh:
```diff
@@ -0,0 +1,55 @@
+
+copyDbToUserDir() {
+  if [ -d /root/.evcc ] && [ ! -f /root/.evcc/.copiedToEvccUser ]; then
+    touch /var/lib/openhab2/.copiedToEvccUser
```
wofür brauchen wir das?

andig replied:
also v.a. unter dem Key openhab? Wäre es nicht besser, die Datei zu verschieben?

pauxus replied:
Das openhab war natürlich ein copy/paste Fehler, sorry. Verschieben ist immer so eine Sache. Falls was schief geht, ist die Datei weg und wir brauchen weitere logik in post-install abort.. So bleibt das Original erhalten und ein revert zur vorherigen Version macht keine Probleme.

Das Flag dient dazu, zu markieren, dass wir die DB bereits kopiert haben. Damit sparen wir uns Unterscheidungen, welches genau die letzte Version war. Wenn es eine DB unter Root gibt UND die nicht als schon kopiert markiert ist, kopier sie rüber. Ist sie schon kopiert worden, kann sie einfach ignoriert werden.

andig commented:
Ich halte das für eine gute Idee, kann aber nicht beurteilen, ob der PR korrekt ist. Können wir einen unabhängigen Review dafür bekommen?

pauxus committed:
Fix copy/paste error, touch only after successful copy

andig commented:
@goebelmeier ist so nett und wird das für mich testen.

pauxus commented:
Bitte die korrigierte Fassung, mein Fix (copy/paste error) war unsinnig, hab es korrigiert.

goebelmeier commented on packaging/scripts/preinstall.sh:
```diff
@@ -0,0 +1,55 @@
+      addgroup --system "$EVCC_GROUP" --quiet
+    fi
+    if ! getent passwd "$EVCC_USER" > /dev/null 2>&1 ; then
+      adduser --quiet --system --ingroup "$EVCC_GROUP" --no-create-home \
```
@pauxus --no-create-home sorgt noch für Issues. Der Folder wird nirgends angelegt und ich bekomme bei einer frischen Installation jetzt:
```
Oct 30 12:38:06 node01 evcc[2069]: [main  ] INFO 2022/10/30 12:38:06 evcc 0.105.2-next
Oct 30 12:38:06 node01 evcc[2069]: [main  ] INFO 2022/10/30 12:38:06 missing config file - switching into demo mode
Oct 30 12:38:06 node01 evcc[2069]: [main  ] INFO 2022/10/30 12:38:06 listening at :7070
Oct 30 12:38:06 node01 evcc[2069]: [main  ] FATAL 2022/10/30 12:38:06 mkdir /var/lib/evcc: permission denied
Oct 30 12:38:06 node01 evcc[2069]: [main  ] FATAL 2022/10/30 12:38:06 will attempt restart in: 5m0s
```

goebelmeier commented on packaging/scripts/preinstall.sh:
```diff
@@ -0,0 +1,55 @@
+      deb-systemd-invoke stop evcc.service >/dev/null || true
+      touch ${RESTART_FLAG_FILE}
+    fi
+    /bin/cp -Rp /root/.evcc/evcc.db "$EVCC_HOME"
```
Schlägt fehl, weil der evcc User mit --no-create-home angelegt wurde. Daher wird /root/.evcc/evcc.db nach /var/lib/evcc kopiert und nicht nach /var/lib/evcc/evcc.db.

```
pi@node01:~ $ cat /var/lib/evcc
ich bin alt
pi@node01:~ $ sudo cat /root/.evcc/evcc.db
ich bin alt
pi@node01:~ $ cat /var/lib/evcc
ich bin alt
pi@node01:~ $ ls -la /var/lib/evcc
-rw-r--r-- 1 root root 13 Oct 30 12:37 /var/lib/evcc
```

pauxus replied:
Danke fürs testen. Ich hatte ursprünglich die Reihenfolge umgekehrt und bei meinen Tests das Verzeichnis natürlich schon angelegt... korrigiere ich

goebelmeier reviewed (commented):
`adduser` should not be called with `--no-create-home` from my point of view.

pauxus committed:
don't skip folder creation

pauxus commented:
@goebelmeier Thanks for testing, I fixed the home dir creation - this might still leave open some very special cornercases (/var/lib/evcc exists but belongs to another user), but I think we are safe to assume someone in that scenario being able to resolve that manually.

andig commented:
Fehlt hier noch was? Sonst wäre jetzt nach Release der richtige Zeitpunkt zum mergen.

/cc @premultiply @VolkerK62

goebelmeier commented:
Gegentesten könnte ich das noch, weiß aber nicht ob ich das morgen unter kriege

pauxus commented:
Wie schaut es aus? Fehlt noch was?

andig commented:
@pauxus Zeit das fertig zu machen ;)

Aktuell liegt die db ja unter `~/.evcc/evcc.db`. Passt das dann immer noch? Du verschiebst sie doch nach `~/evcc`- da würde sie ja dann nicht mehr gefunden, oder funktioniert das bei Dir?

Also wäre die Fragen, wie wir dann wieder an die richtige Location kommen.

pauxus commented:
@andig Ich verschiebe sie nach /var/lib/evcc (also ~evcc statt ~evcc/.evcc), und natürlich habe ich aus irgendeinem Grund vergessen, die angepasste Config mit zu commiten. Argh. Neues Jahr, neues Glück. Ich rebase dann auch gleich nochmal auf den letzten Release.

pauxus commented:
So, nochmal alles durchgesetzt. Funktioniert auch soweit alles, Service läuft mit dem richtigen Nutzer, DB ist kopiert.

Einen Cornercase gibt es, über den ich bei meinem letzten Test gestolpert bin: ich hatte manuell bei mir einen anderen Nutzer über `evcc.service.d/` konfiguriert. Das beachtet das Skript natürlich nicht. Theoretisch könnten wir hier noch einen Schritt weitergehen und den Nutzer, unter dem evcc aktuell läuft bestimmen.

pauxus commented:
Allerdings ist da ein Gedankenfehler drin, muss ich gleich noch korrigieren

pauxus committed:
explicitly override the sqlite database in the unit

Don't change the default location, we only want this for the systemd service

pauxus commented:
Macht natürlich keinen Sinn, den Pfad zur DB global zu überschreiben, sondern nur für die Systemd Unit, ich habe es jetzt explizit im Unitfile gesetzt. Ich werde noch einen Sanity Check was geänderte Nutzer oder doppelte DBs angeht einbauen

andig commented:
> Das beachtet das Skript natürlich nicht.

Das wär mir egal. Wer das kann kann auch die DB wieder finden.

pauxus committed:
More safeguards

- Copy database even if run as non root
- check if database already exists before copying, if so fail
- Fail if existing homedir could not be moved instead of warn only
- Warn if there are systemd overrides, in that case, service is not restarted

pauxus commented:
Hab jetzt noch ein paar Sanity checks dazugepackt, ich denke damit sind alle Szenarien erschlagen

pauxus commented on packaging/init/evcc.service:
```diff
@@ -10,9 +10,12 @@ StartLimitIntervalSec=10
 Restart=always
 RestartSec=10
 
+User=evcc
+Group=evcc
```
Good point - this would simply the script. I will give it a spin tomorrow.

andig replied:
I have no idea what you‘re talking about. Staying tuned :)

pauxus replied:
I did again look into it (was almost tomorrow...) and would rather not use `DynamicUser`. The directories would be created inside /var/lib/private/evcc and symlinked to /var/lib/evcc which introduces unnecessary complexity. Also it would no make the script much simpler. We still would need to create the directory, we still need to copy the database and we would still need to chown (systemd only chowns the containing files if the owner of the directory is wrong, so this might lead to effects, if the user/directory already exists).

The only thing we could do would be to set `StateDirectory` and use the `STATE_DIRECTORY` env variable in the call, but I don't think this would help very much.

mweinelt replied:
Yeah, if you have state to migrate in post-install that gets hairy fast. Most of the issues you describe seem to originate from a migration story.

> if the user/directory already exists

You would need to delete the persistent user.

> we would still need to chown (systemd only chowns the containing files if the owner of the directory is wrong

After removing the persistent user the owner/group would certainly mismatch.

> The directories would be created inside /var/lib/private/evcc and symlinked to /var/lib/evcc which introduces unnecessary complexity.

I think the complexity you see is that it only gets created, when the unit is first started. But as you noticed you can create it yourself and I think that is reasonable given that you want to migrate old state.

I think systemd is a very neat tool to manage a service's lifetime. Find below my personal systemd unit running on NixOS. It goes beyond the scope of this PR, especially by including hardening options, but maybe you'll find it interesting or even useful.

```ini
# /etc/systemd/system/evcc.service
[Unit]
After=network-online.target mosquitto.target mosquitto.service

[Service]
Environment="HOME=/var/lib/evcc"
Environment="LOCALE_ARCHIVE=/nix/store/wqzfxm6fni76h5w9m17q7zkl7h7isk96-glibc-locales-2.35-163/lib/locale/locale-archive"
Environment="PATH=/nix/store/57xv61c5zi8pphjbcwxxjlgc34p61ic9-glibc-2.35-163-bin/bin:/nix/store/a7gvj343m05j2s32xcnwr35v31ynlypr-coreutils-9.1/bin:/nix/store/mydc6f4k2z73xlcz7ilif3v2lcaiqv>
Environment="TZDIR=/nix/store/i16lgq16av602nfyws3ps8dd9yj36dwh-tzdata-2022f/share/zoneinfo"

CapabilityBoundingSet=
DeviceAllow=char-ttyUSB
DevicePolicy=closed
DynamicUser=true
ExecStart=/nix/store/qc0vs3by55v0daf7ii854d56kz3nwsf0-evcc-0.109.2/bin/evcc --config /nix/store/fzbldlqmx5ambi3hmf3jhagb3nvc33qp-evcc.yml 
LockPersonality=true
MemoryDenyWriteExecute=true
PrivateTmp=true
PrivateUsers=true
ProcSubset=pid
ProtectClock=true
ProtectControlGroups=true
ProtectHome=true
ProtectHostname=true
ProtectKernelLogs=true
ProtectKernelModules=true
ProtectKernelTunables=true
ProtectProc=invisible
RestrictAddressFamilies=AF_INET
RestrictAddressFamilies=AF_INET6
RestrictAddressFamilies=AF_UNIX
RestrictNamespaces=true
RestrictRealtime=true
StateDirectory=evcc
SystemCallArchitectures=native
SystemCallFilter=@system-service
SystemCallFilter=~@privileged
UMask=0077
User=evcc
WorkingDirectory=/var/lib/evcc
```

pauxus replied:
> You would need to delete the persistent user.

Since DynamicUser CAN work with preexisting users, this can get fishy rather. If you have a user (log reader or similar) that is part of the group evcc and that group is deleted, you will have unexpected consequences. I would suggest sticking with fixed users. You can always use your really strict (in a positive way, you are correct, I find it one of the most hardened unit files that I have seen so far) unit configuration in an override file.

pauxus replied:
@andig @mweinelt Wie sieht es aus? ich würde es so lassen...

andig commented:
@goebelmeier @premultiply @mweinelt last call- habt ihr noch Anmerkungen? Mir erscheint das sonst vollständig und hoffentlich auch frei von Nebenwirkungen und ich würde es ins Nightly mergen?

premultiply commented:
LGTM

goebelmeier commented:
Es sieht mir gut aus, ich hab relativ viel Fuzzy Testing gemacht und es ist nicht zerbrochen. Also könnte man IMHO reinnehmen

andig commented:
Und natürlich Dank an @pauxus

VolkerK62 commented:
>@VolkerK62 im apt Package wandert die SQlite Datenbank jetzt nach https://github.com/evcc-io/evcc/pull/4901/files#diff-fc00ac250b8d002d87e12d46f86302b54999b140a3048acb82f91de09df794e7R13

Ich habe die Kommentierungen gelesen und nix verstanden.
Muss man bzgl. des Umzugs der DB als normaler Nutzer irgendwie tätig werden?
In der Doku habe ich keinen Verweis auf die DB gefunden, da ist also nix zu tun.

andig commented:
Ne, die sollte per Skript umziehen, liegt aber danach am neuen Ort. Im Logfile sagt evcc, welche verwendet wird

VolkerK62 commented:
Sollte dieser PR in dem Nightly von vorhin (wegen freeze) enthalten sein?

andig commented:
This error has just popped up in the wild:

hatte grade von 0.111.? auf 0.112.5 aktualisiert.
hab jetzt diese Fehler:

```
Feb  8 09:50:26 homeserver evcc[16875]: [lp-1  ] INFO 2023/02/08 09:50:26 start charging ->
Feb  8 09:50:26 homeserver evcc[16875]: [db    ] ERROR 2023/02/08 09:50:26 persist: attempt to write a readonly database (8)
Feb  8 10:28:05 homeserver evcc[16875]: [lp-1  ] INFO 2023/02/08 10:28:05 stop charging <-
Feb  8 10:28:05 homeserver evcc[16875]: [db    ] ERROR 2023/02/08 10:28:05 persist: attempt to write a readonly database (8)
Feb  8 10:28:20 homeserver evcc[16875]: [lp-3  ] INFO 2023/02/08 10:28:20 start charging ->
Feb  8 10:28:20 homeserver evcc[16875]: [db    ] ERROR 2023/02/08 10:28:20 persist: attempt to write a readonly database (8)
Feb  8 10:28:30 homeserver evcc[16875]: [db    ] ERROR 2023/02/08 10:28:30 persist: attempt to write a readonly database (8)
Feb  8 10:28:30 homeserver evcc[16875]: [main  ] ERROR 2023/02/08 10:28:30 cannot save settings: attempt to write a readonly database (1544)
Feb  8 10:28:30 homeserver evcc[16875]: [db    ] ERROR 2023/02/08 10:28:30 persist: attempt to write a readonly database (1544)
```

nach dem ändern "chown evcc:evcc /bla/bla" vom Verzeichnis in dem die SQlite db ist und die db file - läuft alles wieder rund.

-> is that something we need to fix in the packaging?

pauxus commented:
The change was not part of 112.5. You reverted beforehand and revert-reverted afterwards. Need more details:

- which user was used (before/after)?
- Where is bla/bla?
- is this actually a Debian/Ubuntu system

Chris591 commented:
I guess it was root
at least the folder and file was owned by root
bla/bla was the default location `/var/lib/evcc/evcc.db`
OS:
```
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

Chris591 commented:
This was the error or warning I saw during the update.
```
Vorbereitung zum Entpacken von .../evcc_0.112.5+1675824468_armhf.deb ...
adduser: Warnung: Das Home-Verzeichnis ?/var/lib/evcc? geh?rt nicht dem Benutzer, den Sie gerade anlegen.
Entpacken von evcc (0.112.5+1675824468) ?ber (0.111.1+1675414714) ...
evcc (0.112.5+1675824468) wird eingerichtet ...
```

pauxus commented:
Ok, this is a corner case. We could add a chown after the creation of the user, but I don't know whether this is really necessary. 

I assume you created the folder manually? Or do you have any overrides in your systemd module (like stateDirectory)?

Chris591 commented:
No didn't created it but using the nightly and added it to the config file a while ago. Not sure why I had to do it but was related to not saving the chargelogs persistent as I remember correctly.

Chris591 commented:
This is the error on todays update:
```
Vorbereitung zum Entpacken von .../evcc_0.112.5+1675930264_armhf.deb ...
Warning: evcc's home directory is incorrect (116)
but can't be changed because another process (31452) is using it.
Stop offending process(es), then restart installation
dpkg: Fehler beim Bearbeiten des Archivs /var/cache/apt/archives/evcc_0.112.5+1675930264_armhf.deb (--unpack):
 ?neues evcc-Skript des Paketes pre-installation?-Unterprozess gab den Fehlerwert 1 zur?ck
Fehler traten auf beim Bearbeiten von:
 /var/cache/apt/archives/evcc_0.112.5+1675930264_armhf.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

andig commented:
Let's continue in https://github.com/evcc-io/evcc/issues/6162, seems this is more serious.

DerAndereAndi commented:
There is another issue caused by this, which needs to be addressed: https://github.com/evcc-io/evcc/discussions/6140#discussioncomment-4906057

`/dev/tty...` devices can not be accessed by the `evcc` user.
